packages/language/src/preprocessor/instruction-interpreter.ts

```diff
@@ -319,47 +361,241 @@ function runInstruction(
+  node: inst.InstructionNode,
 ): inst.InstructionNode | undefined {
   let condition = true;
   if (instruction.doType2) {
-    const type2 = instruction.doType2;
-    if (type2.until) {
-      const untilConditionValue = evaluateExpression(type2.until, context);
-      if (!isScalarValue(untilConditionValue)) {
-        // Condition cannot be evaluated, don't run the instruction
-        condition = false;
-      } else {
-        // If UNTIL is specified, we don't run the instruction if it evaluates to true
-        condition &&= !valueToBool(untilConditionValue);
-      }
-    }
-    if (type2.while) {
-      const whileConditionValue = evaluateExpression(type2.while, context);
-      if (!isScalarValue(whileConditionValue)) {
-        // Condition cannot be evaluated, don't run the instruction
-        condition = false;
-      } else {
-        // If WHILE is specified, we only run the instruction if it evaluates to true
-        condition &&= valueToBool(whileConditionValue);
-      }
-    }
-  }
-  if (instruction.doType3) {
-    throw new Error("DoType3 instructions are not implemented yet!");
+    condition = runDoType2Instruction(instruction.doType2, context);
+  } else if (instruction.doType3) {
+    condition = runDoType3Instruction(instruction.doType3, context, node);
   }
   if (condition) {
     return instruction.content;
   }
   return undefined;
 }
 
+function runDoType2Instruction(
+  doType2: inst.DoType2Instruction,
+  context: InterpreterContext,
+): boolean {
+  let condition = true;
+  if (doType2.until) {
+    const untilConditionValue = evaluateExpression(doType2.until, context);
+    if (!isScalarValue(untilConditionValue)) {
+      // Condition cannot be evaluated, don't run the instruction
+      condition = false;
+    } else {
+      // If UNTIL is specified, we don't run the instruction if it evaluates to true
+      condition &&= !valueToBool(untilConditionValue);
+    }
+  }
+  if (doType2.while) {
+    const whileConditionValue = evaluateExpression(doType2.while, context);
+    if (!isScalarValue(whileConditionValue)) {
+      // Condition cannot be evaluated, don't run the instruction
+      condition = false;
+    } else {
+      // If WHILE is specified, we only run the instruction if it evaluates to true
+      condition &&= valueToBool(whileConditionValue);
+    }
+  }
+  return condition;
+}
+
+function runDoType3Instruction(
+  doType3: inst.DoType3Instruction,
+  context: InterpreterContext,
+  node: inst.InstructionNode,
+): boolean {
+  let condition = true;
+
+  // Get current do-type3 state
+  const doType3Context = context.doType3.get(node);
+
+  if (!doType3Context) {
+    // Store all expressions that must be evaluated at the start of the loop
+    condition &&= runDoType3Initialization(doType3, context, node);
+  } else {
+    // Subsequent iterations
+    condition &&= runDoType3UntilCheck(doType3, context);
+    condition &&= runDoType3VariableUpdate(doType3, context, node);
```

Comment:
Suggestion: Shouldn't we pass the `doType3Context` instead of the node here? Otherwise we use `doType3.get(node)` twice.
```suggestion
    condition &&= runDoType3VariableUpdate(doType3, context, doType3Context);
```

Reply:
Good catch, Sir! That makes absolutely sense.